File: app/src/main/java/com/nathaniel/motus/cavevin/view/MyRecyclerViewAdapter.java
package com.nathaniel.motus.cavevin.view;

import android.app.Activity;
import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.nathaniel.motus.cavevin.R;
import com.nathaniel.motus.cavevin.controller.CellarPictureUtils;
import com.nathaniel.motus.cavevin.controller.CellarStorageUtils;
import com.nathaniel.motus.cavevin.controller.MainActivity;
import com.nathaniel.motus.cavevin.model.Bottle;
import com.nathaniel.motus.cavevin.model.Cell;
import com.nathaniel.motus.cavevin.model.Cellar;

import static com.nathaniel.motus.cavevin.R.*;

public class MyRecyclerViewAdapter extends RecyclerView.Adapter<MyViewHolder>{

    private Cellar mCellar;
    private Context mContext;
    private Activity mActivity;
    private onItemClickedListener mCallback;

    public MyRecyclerViewAdapter(Cellar cellar, Context context, Activity activity) {
        mCellar = cellar;
        mContext=context;
        mActivity=activity;
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView= LayoutInflater.from(parent.getContext()).inflate(layout.recyclerview_cellar_row,parent,false);
        MyViewHolder mViewHolder=new MyViewHolder(itemView);
        return mViewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull final MyViewHolder holder, final int position) {

//                Get the current cellar position before callback (position is filtered cellar position)
        final int actualPosition=Cellar.getCellarPool().get(MainActivity.getCurrentCellarIndex()).getCellList().indexOf(mCellar.getCellList().get(position));

        //calculate width of multilines textviews
        int typeWidthPx= (int) mActivity.getResources().getDimension(dimen.recyclerview_cellar_row_type_width);
        int photoImageWidthPx= (int) mActivity.getResources().getDimension(dimen.recyclerview_cellar_row_photo_width);
        int vintageWidthPx= (int) mActivity.getResources().getDimension(dimen.recyclerview_cellar_row_vintage_width);
        int buttonsWidthPx= (int) mActivity.getResources().getDimension(dimen.recyclerview_cellar_row_buttons_width);

        int appellationWidth=CellarPictureUtils.getDisplayWidthPx(mContext)-typeWidthPx-photoImageWidthPx-vintageWidthPx-buttonsWidthPx;
        int domainWidth=CellarPictureUtils.getDisplayWidthPx(mContext)-typeWidthPx-photoImageWidthPx-buttonsWidthPx;
        int cuveeWidth=CellarPictureUtils.getDisplayWidthPx(mContext)-typeWidthPx-photoImageWidthPx-buttonsWidthPx;

        holder.appellation.getLayoutParams().width=appellationWidth;
        holder.domain.getLayoutParams().width=domainWidth;
        holder.cuvee.getLayoutParams().width=cuveeWidth;


        final Cell cell=mCellar.getCellList().get(position);
        Bottle bottle=cell.getBottle();
        String color=bottle.getType();
        switch (color){
            case "1":
                holder.type.setBackgroundColor(mContext.getResources().getColor(R.color.whiteWine));
                break;
            case "2":
                holder.type.setBackgroundColor(mContext.getResources().getColor(R.color.pinkWine));
                break;
            default:
                holder.type.setBackgroundColor(mContext.getResources().getColor(R.color.redWine));
        }
        holder.appellation.setText(bottle.getAppellation());
        holder.domain.setText(bottle.getDomain());
        holder.cuvee.setText(bottle.getCuvee());
        holder.vintage.setText(bottle.getVintage());
        holder.bottleName.setText(bottle.getBottleName());
        holder.capacity.setText(Float.toString(bottle.getCapacity())+" L");
        holder.quantity.setText(Integer.toString(cell.getStock()));
        holder.origin.setText(cell.getOrigin());
        holder.bottleComment.setText(bottle.getBottleComment());
        holder.cellarComment.setText(cell.getCellComment());

        //load photo image
        String photoPathName=bottle.getPhotoPathName();
        if(photoPathName.compareTo("")!=0)
//            holder.photoImage.setImageBitmap(CellarStorageUtils.getBitmapFromInternalStorage(mContext.getFilesDir(), mContext.getResources().getString(string.photo_folder_name),photoPathName));
            holder.photoImage.setImageBitmap(CellarStorageUtils.decodeSampledBitmapFromFile(mContext.getFilesDir(),mContext.getResources().getString(string.photo_folder_name),
                    photoPathName,(int)mContext.getResources().getDimension(dimen.recyclerview_cellar_row_photo_width),(int)mContext.getResources().getDimension(dimen.recyclerview_cellar_row_photo_height)));
        else
            holder.photoImage.setImageDrawable(mContext.getResources().getDrawable(R.drawable.photo_frame));

        createCallBackToParentActivity();
        holder.buttonEdit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mCallback.onItemClicked(v,actualPosition);
            }
        });

        holder.buttonPlus.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Cell cell=Cellar.getCellarPool().get(MainActivity.getCurrentCellarIndex()).getCellList().get(actualPosition);
                cell.setStock(cell.getStock()+1);
                holder.quantity.setText(Integer.toString(cell.getStock()));
            }
        });

        holder.buttonMinus.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Cell cell=Cellar.getCellarPool().get(MainActivity.getCurrentCellarIndex()).getCellList().get(actualPosition);
                if (cell.getStock()>0) cell.setStock(cell.getStock()-1);
                holder.quantity.setText(Integer.toString(cell.getStock()));
            }
        });

        holder.buttonCollapse.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (holder.originTitle.getVisibility()==View.VISIBLE) {
                    holder.originTitle.setVisibility(View.GONE);
                    holder.origin.setVisibility(View.GONE);
                    holder.bottleCommentTitle.setVisibility(View.GONE);
                    holder.bottleComment.setVisibility(View.GONE);
                    holder.cellarCommentTitle.setVisibility(View.GONE);
                    holder.cellarComment.setVisibility(View.GONE);
                    holder.buttonEdit.setVisibility(View.GONE);
                    holder.buttonCollapse.setImageResource(drawable.baseline_keyboard_arrow_down_white_18dp);
                }else{
                    holder.originTitle.setVisibility(View.VISIBLE);
                    holder.origin.setVisibility(View.VISIBLE);
                    holder.bottleCommentTitle.setVisibility(View.VISIBLE);
                    holder.bottleComment.setVisibility(View.VISIBLE);
                    holder.cellarCommentTitle.setVisibility(View.VISIBLE);
                    holder.cellarComment.setVisibility(View.VISIBLE);
                    holder.buttonEdit.setVisibility(View.VISIBLE);
                    holder.buttonCollapse.setImageResource(drawable.baseline_keyboard_arrow_up_white_18dp);

                }
            }
        });
    }



    @Override
    public int getItemCount() {
        return mCellar.getCellList().size();
    }

//    **********************************************************************************************
//    Callback subs
//    **********************************************************************************************

    public interface onItemClickedListener{
        public void onItemClicked(View view,int position);
    }

    private void createCallBackToParentActivity(){
        try{
            mCallback=(onItemClickedListener)mActivity;
        }catch (ClassCastException e){
            throw new ClassCastException(e.toString()+" must implement onItemClickedListener");
        }
    }
}

File: app/src/main/java/com/nathaniel/motus/cavevin/view/MyRecyclerViewAdapter.kt
package com.nathaniel.motus.cavevin.view

import com.nathaniel.motus.cavevin.controller.CellarPictureUtils.getDisplayWidthPx
import com.nathaniel.motus.cavevin.controller.CellarStorageUtils.getBitmapFromInternalStorage
import androidx.recyclerview.widget.RecyclerView
import com.nathaniel.motus.cavevin.R
import android.view.LayoutInflater
import android.view.ViewGroup
import com.nathaniel.motus.cavevin.model.Cellar
import com.nathaniel.motus.cavevin.controller.MainActivity
import android.app.Activity
import android.content.Context
import android.view.View
import com.nathaniel.motus.cavevin.R.layout
import com.nathaniel.motus.cavevin.R.dimen
import com.nathaniel.motus.cavevin.R.string
import com.nathaniel.motus.cavevin.R.drawable
import com.nathaniel.motus.cavevin.model.Cell
import java.lang.ClassCastException

class MyRecyclerViewAdapter(
    private val mCellar: Cellar,
    private val mContext: Context,
    private val mActivity: Activity?
) : RecyclerView.Adapter<MyViewHolder>() {
    private var mCallback: onItemClickedListener? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val itemView = LayoutInflater.from(parent.context)
            .inflate(layout.recyclerview_cellar_row, parent, false)
        return MyViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {

//                Get the current cellar position before callback (position is filtered cellar position)
        val actualPosition: Int =
            Cellar.cellarPool[MainActivity.currentCellarIndex].cellList.indexOf(
                mCellar.cellList[position]
            )

        //calculate width of multilines textviews
        val typeWidthPx =
            mActivity!!.resources.getDimension(dimen.recyclerview_cellar_row_type_width)
                .toInt()
        val photoImageWidthPx =
            mActivity.resources.getDimension(dimen.recyclerview_cellar_row_photo_width)
                .toInt()
        val vintageWidthPx =
            mActivity.resources.getDimension(dimen.recyclerview_cellar_row_vintage_width)
                .toInt()
        val buttonsWidthPx =
            mActivity.resources.getDimension(dimen.recyclerview_cellar_row_buttons_width)
                .toInt()
        val appellationWidth =
            getDisplayWidthPx(mContext) - typeWidthPx - photoImageWidthPx - vintageWidthPx - buttonsWidthPx
        val domainWidth =
            getDisplayWidthPx(mContext) - typeWidthPx - photoImageWidthPx - buttonsWidthPx
        val cuveeWidth =
            getDisplayWidthPx(mContext) - typeWidthPx - photoImageWidthPx - buttonsWidthPx
        holder.appellation.layoutParams.width = appellationWidth
        holder.domain.layoutParams.width = domainWidth
        holder.cuvee.layoutParams.width = cuveeWidth
        val cell = mCellar.cellList[position]!!
        val bottle = cell.bottle
        val color = bottle.type
        when (color) {
            "1" -> holder.type.setBackgroundColor(mContext.resources.getColor(R.color.whiteWine))
            "2" -> holder.type.setBackgroundColor(mContext.resources.getColor(R.color.pinkWine))
            else -> holder.type.setBackgroundColor(mContext.resources.getColor(R.color.redWine))
        }
        holder.appellation.text = bottle.appellation
        holder.domain.text = bottle.domain
        holder.cuvee.text = bottle.cuvee
        holder.vintage.text = bottle.vintage
        holder.bottleName.text = bottle.bottleName
        holder.capacity.text = java.lang.Double.toString(bottle.capacity) + " L"
        holder.quantity.text = Integer.toString(cell.stock)
        holder.origin.text = cell.origin
        holder.bottleComment.text = bottle.bottleComment
        holder.cellarComment.text = cell.cellComment

        //load photo image
        val photoName = bottle.photoName
        if (photoName.compareTo("") != 0) holder.photoImage.setImageBitmap(
            getBitmapFromInternalStorage(
                mContext.filesDir,
                mContext.resources.getString(string.photo_folder_name),
                photoName + mContext.resources.getString(string.photo_thumbnail_suffix)
            )
        ) else holder.photoImage.setImageDrawable(mContext.resources.getDrawable(drawable.photo_frame))

        //Expand or collapse, according to flag isExpanded
        if (mCellar.cellList[position]!!.isExpanded) expandCard(holder) else collapseCard(holder)
        createCallBackToParentActivity()

        //Click listeners
        holder.buttonEdit.setOnClickListener { v -> mCallback!!.onItemClicked(v, actualPosition) }
        holder.buttonPlus.setOnClickListener {
            val cell: Cell =
                Cellar.cellarPool[MainActivity.currentCellarIndex].cellList[actualPosition]!!
            cell.stock = cell.stock + 1
            holder.quantity.text = Integer.toString(cell.stock)
        }
        holder.buttonMinus.setOnClickListener {
            val cell: Cell =
                Cellar.cellarPool[MainActivity.currentCellarIndex].cellList[actualPosition]!!
            if (cell.stock > 0) cell.stock = cell.stock - 1
            holder.quantity.text = Integer.toString(cell.stock)
        }
        holder.photoImage.setOnClickListener { v -> mCallback!!.onItemClicked(v, actualPosition) }
        holder.buttonCollapse.setOnClickListener {
            if (mCellar.cellList[position]!!.isExpanded) {
                collapseCard(holder)
                mCellar.cellList[position]!!.isExpanded = false
            } else {
                expandCard(holder)
                mCellar.cellList[position]!!.isExpanded = true
            }
        }
    }

    override fun getItemCount(): Int {
        return mCellar.cellList.size
    }

    //    **********************************************************************************************
    //    Callback subs
    //    **********************************************************************************************
    interface onItemClickedListener {
        fun onItemClicked(view: View?, position: Int)
    }

    private fun createCallBackToParentActivity() {
        mCallback = try {
            mActivity as onItemClickedListener?
        } catch (e: ClassCastException) {
            throw ClassCastException("$e must implement onItemClickedListener")
        }
    }

    //    **********************************************************************************************
    //    Display subs
    //    **********************************************************************************************
    private fun collapseCard(holder: MyViewHolder) {
        //Collapse a card
        holder.originTitle.visibility = View.GONE
        holder.origin.visibility = View.GONE
        holder.bottleCommentTitle.visibility = View.GONE
        holder.bottleComment.visibility = View.GONE
        holder.cellarCommentTitle.visibility = View.GONE
        holder.cellarComment.visibility = View.GONE
        holder.buttonEdit.visibility = View.GONE
        holder.buttonCollapse.setImageResource(drawable.baseline_keyboard_arrow_down_white_18dp)
    }

    private fun expandCard(holder: MyViewHolder) {
        //Expand a card
        holder.originTitle.visibility = View.VISIBLE
        holder.origin.visibility = View.VISIBLE
        holder.bottleCommentTitle.visibility = View.VISIBLE
        holder.bottleComment.visibility = View.VISIBLE
        holder.cellarCommentTitle.visibility = View.VISIBLE
        holder.cellarComment.visibility = View.VISIBLE
        holder.buttonEdit.visibility = View.VISIBLE
        holder.buttonCollapse.setImageResource(drawable.baseline_keyboard_arrow_up_white_18dp)
    }
}